cr-usecases/src/main/java/com/thefuntasty/mvvm/crusecases/Result.kt

```diff
@@ -0,0 +1,273 @@
+package com.thefuntasty.mvvm.crusecases
+
+import kotlinx.coroutines.CancellationException
+
+/**
+ * A discriminated union that encapsulates successful outcome either with [Success] or with [Error]
+ *
+ * This class is adapted to kotlin coroutines so if [Error] contains [CancellationException] then this error is rethrown.
+ */
+sealed class Result<out VALUE : Any?> {
+    override fun toString(): String {
+        return when (this) {
+            is Success -> "Success($value)"
+            is Error -> "Error(${error})"
+        }
+    }
+
+    operator fun component1(): VALUE? = if (this is Success) value else null
+
+    operator fun component2(): Throwable? = if (this is Error) error else null
+}
+
+/**
+ * Represents a successful [Result], containing a [value].
+ */
+class Success<VALUE : Any?>(val value: VALUE) : Result<VALUE>() {
+
+    override fun equals(other: Any?): Boolean {
```

Comment:
<table data-meta="generated_by_danger-pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/thefuntasty/mvvm-android/blob/e74b43dc1369be07c51adbdf506d5277c7cb50ed/cr-usecases/src/main/java/com/thefuntasty/mvvm/crusecases/Result.kt#L28"></span>Function equals has 5 return statements which exceeds the limit of 4.</td>
    </tr>
  </tbody>
</table>